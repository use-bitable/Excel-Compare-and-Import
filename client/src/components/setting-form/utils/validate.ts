import type { fieldMap } from "@/types/types"
import { Warn, autoFields } from "@/utils"

/**
 * Validate if index is empty or has no excel_field, if yes, return false, else return true
 *
 * @param index
 * @returns
 */
export function validateIndex(index: fieldMap[]) {
  if (index.length === 0) {
    Warn({
      title: "noIndex",
      message: "noIndex",
      notice: true,
      noticeParams: {
        text: "message.noIndex",
      },
    })
    return false
  }
  const noExcelField = index.filter((i) => !i.excel_field)
  if (noExcelField.length) {
    const names = noExcelField.map((i) => i.field.name).join(", ")
    Warn({
      title: "noExcelField",
      message: names,
      notice: true,
      noticeParams: {
        text: "message.indexNoExcelField",
        params: {
          fields: names,
        },
      },
    })
    return false
  }
  return true
}

/**
 * Validate if index contains auto field, if yes, return true, else return false
 *
 * @param index
 * @returns
 */
export function validateIndexAuto(index: fieldMap[]) {
  const auto = index.filter((i) => autoFields.includes(i.field.type))
  if (auto.length) {
    const names = auto.map((i) => i.field.name).join(", ")
    Warn({
      title: "autoFieldInIndex",
      message: "autoFieldInIndex: " + names,
      notice: true,
      noticeParams: {
        text: "message.autoFieldInIndex",
        params: {
          fields: names,
        },
      },
    })
    return true
  }
  return false
}
