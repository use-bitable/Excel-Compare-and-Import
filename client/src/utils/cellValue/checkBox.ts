import { type ICheckBoxField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator } from "./cell"
import type { fieldMap } from "@/types/types"
export const defaultBoolValue = {
  true: ["是", "True", "true", "TRUE", "1", "☑️"],
  false: ["否", "False", "false", "FALSE", "0", ""],
}

async function normalization(value: string, config?: fieldMap["config"]) {
  const { boolValue = defaultBoolValue } = config || {}
  return boolValue.true.includes(value)
}

/**
 * Get checkBox cell
 *
 * @param value
 * @param field
 * @param config
 * @returns
 */
async function checkBox(
  value: string,
  field: ICheckBoxField,
  config?: fieldMap["config"],
) {
  const v = await normalization(value, config)
  return await field.createCell(v)
}

export const CheckBoxTranslator = defineTranslator({
  fieldType: FieldType.Checkbox,
  translate: checkBox,
  normalization,
  name: "CheckBox",
})
