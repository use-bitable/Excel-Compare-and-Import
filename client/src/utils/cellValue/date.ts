import dayjs from "dayjs/esm/index.js"
import customParseFormat from "dayjs/plugin/customParseFormat"
import advancedFormat from "dayjs/plugin/advancedFormat"
import { type IDateTimeField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator } from "./cell"
import type { fieldMap } from "@/types/types"

dayjs.extend(customParseFormat)
dayjs.extend(advancedFormat)

export const dateDefaultFormat = "YYYY/MM/DD"

export async function normalization(
  value: string,
  config?: fieldMap["config"],
) {
  const { format = dateDefaultFormat } = config || {}
  return dayjs(value, format).valueOf()
}

/**
 * Get dateTime cell
 *
 * @param value
 * @param field
 * @param config
 * @returns
 */
async function dateTime(
  value: string,
  field: IDateTimeField,
  config?: fieldMap["config"],
) {
  const v = await normalization(value, config)
  return await field.createCell(v)
}

export const DateTimeTranslator = defineTranslator({
  fieldType: FieldType.DateTime,
  translate: dateTime,
  normalization,
  name: "DateTime",
})
