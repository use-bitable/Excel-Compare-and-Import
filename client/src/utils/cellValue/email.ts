import { type IEmailField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator } from "./cell"

/**
 * Get text cell
 *
 * @param value
 * @param field
 * @returns
 */
async function email(value: string, field: IEmailField) {
  return await field.createCell(value)
}

export const EmailTranslator = defineTranslator({
  fieldType: FieldType.Email,
  translate: email,
  normalization: async (value: string) => value,
  name: "Email",
})
