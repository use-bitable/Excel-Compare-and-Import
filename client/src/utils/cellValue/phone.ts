import { type IPhoneField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator } from "./cell"

/**
 * Get phone cell
 *
 * @param value
 * @param field
 * @returns
 */
async function phone(value: string, field: IPhoneField) {
  return await field.createCell(value)
}

export const PhoneTranslator = defineTranslator({
  fieldType: FieldType.Phone,
  translate: phone,
  normalization: async (value: string) => value,
  name: "Phone",
})
