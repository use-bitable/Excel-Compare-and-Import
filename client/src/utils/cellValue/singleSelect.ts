import { type ISingleSelectField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator, FieldRole } from "./cell"

/**
 * Get singleSelect cell
 * @param field
 * @param value
 * @returns
 */
async function singleSelect(value: string, field: ISingleSelectField) {
  return await field.createCell(value)
}

export const SingleSelectTranslator = defineTranslator({
  fieldType: FieldType.SingleSelect,
  translate: singleSelect,
  normalization: async (value: string) => value,
  name: "SingleSelect",
  roles: [FieldRole.HAS_OPTIONS],
})
