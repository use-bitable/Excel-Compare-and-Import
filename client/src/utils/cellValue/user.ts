import { type IUserField, FieldType } from "@lark-base-open/js-sdk"
import { defineTranslator } from "./cell"
import type { fieldMap } from "@/types/types"

export const userSeparator = ","

interface UserCache {
  multiple: boolean
}

let cache: Record<string, UserCache> = {}

async function setCache(field: IUserField) {
  const multiple = await field.getMultiple()
  cache[field.id] = {
    multiple,
  }
  return {
    multiple,
  }
}

function refreshCache() {
  cache = {}
}

async function normalization(value: string, config?: fieldMap["config"]) {
  const { separator = userSeparator } = config || {}
  return value.split(separator)
}

/**
 * Get user cell
 *
 * @param openId
 * @param field
 * @param config
 * @returns
 */
async function User(
  openId: string,
  field: IUserField,
  config?: fieldMap["config"],
) {
  const { multiple } = cache[field.id] || (await setCache(field))
  const users = await normalization(openId, config)
  return await field.createCell(
    multiple
      ? users.map((u) => ({
          id: u,
        }))
      : { id: users[0] },
  )
}

export const UserTranslator = defineTranslator({
  fieldType: FieldType.User,
  translate: User,
  name: "User",
  normalization,
  refresh: refreshCache,
})
