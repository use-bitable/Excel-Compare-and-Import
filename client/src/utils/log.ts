import { ref } from "vue"

export type LogOptions = {
  prefix: string
  onError?: (msg: Message) => void
  onWarn?: (msg: Message) => void
  onInfo?: (msg: Message) => void
  onLog?: (msg: Message) => void
}

interface NoticeParams {
  text: string
  params?: Record<string, string>
}

export type Message = {
  title?: string
  message: string
  error?: any
  notice?: boolean
  noticeParams?: NoticeParams
}

export type LogItem = {
  type: "info" | "warn" | "error" | "log"
  message: Message
  time: number
  file?: string
  line?: number
}

export const log = ref<LogItem[]>([])

export const options = ref<LogOptions>({
  prefix: "[Excel Compare and Import] ",
})

export function setLogOptions(newOptions: Partial<LogOptions>) {
  options.value = {
    ...options.value,
    ...newOptions,
  }
}

const _console = console

export const useLog =
  <T extends LogItem["type"]>(type: T) =>
  (msg: Message) => {
    const cache = {
      type: type,
      message: msg,
      time: Date.now(),
    }
    log.value.push(cache)
    const action = _console[type] ?? _console.log
    action.apply(void 0, [
      options.value.prefix + msg.title + ": ",
      msg.message,
      msg.error ?? "",
    ])

    if (type === "error") {
      options.value.onError?.(msg)
    }

    if (type === "warn") {
      options.value.onWarn?.(msg)
    }

    if (type === "info") {
      options.value.onInfo?.(msg)
    }

    if (type === "log") {
      options.value.onLog?.(msg)
    }
  }

export const Log = useLog("log")
export const Info = useLog("info")
export const Warn = useLog("warn")
export const Error = useLog("error")
