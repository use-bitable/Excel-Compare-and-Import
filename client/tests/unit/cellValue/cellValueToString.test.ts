import { cellValueToString } from "../../../src/utils/cellValue"
import { describe, it, expect } from "vitest"

describe("cellValueToString test", () => {
  it("IOpenSegment[]", () => {
    const value = [
      {
        type: "text", // 文本类型
        text: "Hello world",
      }, // type: IOpenTextSegment
      {
        type: "url", // URL 类型
        text: "https://feishu.cn",
        link: "https://feishu.cn",
      }, // type: IOpenUrlSegment
      {
        type: "mention", // @人 类型
        mentionType: "User", //
        name: "李健", // 人名
        text: "@李健",
        en_name: "Jian Li",
        token: "ou_xxx", // OpenId，暂未支持
      }, // type: IOpenUserMentionSegment
      {
        type: "mention", // 文档类型
        mentionType: "Bitable", // Bitable 文档
        text: "未命名多维表格",
        token: "basxxxx",
        link: "https://www.feishu.cn/base/basxxxxx",
      }, // type: IOpenDocumentMentionSegment
    ]
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("IOpenSingleSelect", () => {
    const value = {
      id: "option_id",
      text: "option_text",
    }
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("IOpenMultiSelect", () => {
    const value = [
      {
        id: "option_id",
        text: "option_text",
      },
      {
        id: "option_id",
        text: "option_text",
      },
    ]
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("IOpenUser[]", () => {
    const value = [
      {
        name: "@李健", // 人名
        enName: "@JianLi", // 英文名
        email: "lijian@feishu.cn", // 邮箱
        id: "ou_xxx", // OpenId
      },
    ]
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("number", () => {
    const value = 1
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("string", () => {
    const value = "string"
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("IOpenCheckbox", () => {
    const value = true
    expect(cellValueToString(value)).toMatchSnapshot()
  })

  it("IOpenLocation", () => {
    const value = {
      address: "科苑地铁站C出口",
      adname: "南山区",
      cityname: "深圳市",
      full_address: "深圳湾创新科技中心，广东深圳南山科苑地铁站C出口",
      fullAddress: "深圳湾创新科技中心，广东深圳南山科苑地铁站C出口",
      location: "113.946874,22.52835",
      name: "深圳湾创新科技中心",
      pname: "广东省",
    }
    expect(cellValueToString(value)).toMatchSnapshot()
  })
})
