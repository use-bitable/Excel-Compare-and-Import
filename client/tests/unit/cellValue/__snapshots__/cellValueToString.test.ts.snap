// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cellValueToString test > IOpenCheckbox 1`] = `"☑️"`;

exports[`cellValueToString test > IOpenLocation 1`] = `"深圳湾创新科技中心，广东深圳南山科苑地铁站C出口"`;

exports[`cellValueToString test > IOpenMultiSelect 1`] = `"option_text,option_text"`;

exports[`cellValueToString test > IOpenSegment[] 1`] = `"Hello worldhttps://feishu.cn@李健未命名多维表格"`;

exports[`cellValueToString test > IOpenSingleSelect 1`] = `"option_text"`;

exports[`cellValueToString test > IOpenUser[] 1`] = `"ou_xxx"`;

exports[`cellValueToString test > number 1`] = `"1"`;

exports[`cellValueToString test > string 1`] = `undefined`;
